src/target/riscv/riscv-013.c

```diff
@@ -4270,6 +4270,35 @@ read_memory_progbuf(struct target *target, target_addr_t address,
 		read_memory_progbuf_inner_one(target, access, mprven) :
 		read_memory_progbuf_inner(target, access, count, mprven);
 
+	if (result != ERROR_OK) {
+		/* The full read did not succeed, so we will try to read each word individually. */
+		/* This will not be fast, but reading outside actual memory is a special case anyway. */
+		/* It will make the toolchain happier, especially Eclipse Memory View as it reads ahead. */
+		target_addr_t address_i = access.target_address;
+		uint8_t *buffer_i = access.buffer_address;
+
+		for (uint32_t i = 0; i < count; i++, address_i += access.increment, buffer_i += access.element_size) {
+			/* TODO: This is much slower than it needs to be because we end up
+			 * writing the address to read for every word we read. */
+			const struct memory_access_info access_single = {
+			.target_address = address_i,
+			.increment = increment,
+			.buffer_address = buffer_i,
+			.element_size = size,
+			};
+			result = read_memory_progbuf_inner_one(target, access_single, mprven);
+
+			/* The read of a single word failed, so we will just return 0 for that instead */
+			if (result != ERROR_OK) {
```

Comment:
I don't get it. If read fails - the error should be returned.

Reply:
If a section of memory read by openocd starts with an illegal address and ends with a legal address, openocd will directly return an error for the entire memory.The added code can return 0 for memory reading from illegal addresses and return the correct value for memory reading from legal addresses.